### PR TITLE
Rspec tests

### PR DESCRIPTION
**Pull request by rsundar**
Please review the rspec tests for my Tic Tac Toe implementation.

I am working alone since there is no one available in the same time zone as I am for this course.

Please don't hesitate to contact me on slack ( @rsundar ) if any significant changes are required.

Thanks

### CODE REVIEW

File: spec/board_spec.rb [hunk 1]
```diff
@@ -0,0 +1,107 @@
+require 'rspec'
+require '../lib/board.rb'
```
Comment (wathigo) [inline spec/board_spec.rb]: **[OPTIONAL]**
* Let's use `require_relative` instead of `require` in order to run the tests from the root directory.

**Review by wathigo — CHANGES_REQUESTED** [review]
# Changes Required 
## Great work so far :clap:  
### [REQUIRED]
* Let's make sure that all the public methods in player, game, and board classes has been tested. See this :point_down: 
uaw.
Use this template to guide you :point_down: 
```
describe Board do
  describe '#update_board' do
    it '....' do
      # some code here
    end
    it '....' do
      # some code here
    end
  end
end
```
### After implementing the changes, submit another request for review.
**NB:** _All the current tests are okay, let's add some more to ensure all the public methods have been tested. Feel free to reach out on slack for any issue or clarification_

**Happy Coding :computer: 
Simon Wathgo**

File: spec/board_spec.rb (hunk 1 above)
Comment (rsundar) [reply spec/board_spec.rb]: Done!

**Comment by rsundar** [comment]
# Summary of Changes Made

As per the recommendations made by the previous reviewer ( @wathigo ) the following changes have been made:

- Changed the rspec tests in all of the spec files to reflect the recommended template of
  ```
      describe "#method" do
               it "..." do
                    #code here
               end
      end 
``` ```
- Added tests for the ```is_valid?``` method in the ```Game``` class.

Please contact me on Slack if any further changes are required. ( @rsundar )

File: spec/board_spec.rb [hunk 2]
```diff
@@ -0,0 +1,107 @@
+require 'rspec'
```
Comment (maelfosso) [inline spec/board_spec.rb]: Change it with 
```ruby
require 'spec_helper'
```

File: spec/game_spec.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
+require 'rspec'
```
Comment (maelfosso) [inline spec/game_spec.rb]: Change it with 
```ruby
require 'spec_helper'
```

File: spec/player_spec.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
+require 'rspec'
```
Comment (maelfosso) [inline spec/player_spec.rb]: Change it with 
```ruby
require 'spec_helper'
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Change required

Hello @rsundar 

Thanks for the changes, the improvements are awesome.
However, I had a problem when running the test. 
```ruby
$ rspec

An error occurred while loading ./spec/board_spec.rb. - Did you mean?
                    rspec ./spec/game_spec.rb
                    rspec ./spec/player_spec.rb

Failure/Error: require 'rspec'

LoadError:
  cannot load such file -- rspec
# ./spec/board_spec.rb:1:in `<top (required)>'

An error occurred while loading ./spec/game_spec.rb. - Did you mean?
                    rspec ./spec/board_spec.rb
                    rspec ./spec/player_spec.rb

Failure/Error: require 'rspec'

LoadError:
  cannot load such file -- rspec
# ./spec/game_spec.rb:1:in `<top (required)>'

An error occurred while loading ./spec/player_spec.rb. - Did you mean?
                    rspec ./spec/board_spec.rb
                    rspec ./spec/game_spec.rb

Failure/Error: require 'rspec'

LoadError:
  cannot load such file -- rspec
# ./spec/player_spec.rb:1:in `<top (required)>'
No examples found.


Finished in 0.00003 seconds (files took 0.33212 seconds to load)
0 examples, 0 failures, 3 errors occurred outside of examples
```
Please use **spec_helper** file. I suggest the changes to do to solve it

Kindly make that changes and submit a new code review request.
Thanks

**Mael FOSSO**
[Github](https://github.com) | [Twitter](https://twitter.com) | [LinkedIn](https://www.linkedin.com/in/mael-elvis-fosso-650b6346/)

File: spec/board_spec.rb (hunk 2 above)
Comment (rsundar) [reply spec/board_spec.rb]: Done

File: spec/game_spec.rb (hunk 3 above)
Comment (rsundar) [reply spec/game_spec.rb]: done

File: spec/player_spec.rb (hunk 4 above)
Comment (rsundar) [reply spec/player_spec.rb]: done

**Comment by rsundar** [comment]
# Summary of Changes Made

Corrected the error pointed out by @maelfosso 

- Changed all of the ```require 'rspec'``` at the beginning of each of the ```board_spec```,```game_spec```, and ```player_spec``` files in the ```spec``` directory to ```require 'spec_helper'```.

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Good job :+1: 
You did well implementing the suggested changes. I tested your code and am impressed most cases have passed. However, we need to fix a few edge cases before we can merge this one.
Please check my comments and make the requested changes.

### Martin Afani

File: spec/board_spec.rb [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        it "Check if the board is created" do
+            expect(board).to be_a(Board)
+        end
+
+        it "Check if the winning layout is initialized" do
+            expect(board.cells).to eq(winner)
+        end
+
+        it "Check if the board has a winner. Should return true" do
```
Comment (whiz25) [inline spec/board_spec.rb:53]: - [x] Board Winning Board Check if the board has a winner. Should return true

File: spec/board_spec.rb [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            expect(board.winner?).to be true
+        end
+
+        it "The draw? method should return false" do
```
Comment (whiz25) [inline spec/board_spec.rb:57]: - [x] Board Winning Board The draw? method should return false

File: spec/board_spec.rb [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+        it "Check if the drawn board layout is initialized" do
+            expect(board.cells).to eq(draw)
+        end
+
+        it "Check if the number of moves available is 0" do
+            expect(board.moves).to eq(0)
+        end
+
+        it "Check if the game is a draw. Should be true" do
```
Comment (whiz25) [inline spec/board_spec.rb:78]: - [x] Board Drawn game Check if the game is a draw. Should be true

File: spec/board_spec.rb [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            expect(board.draw?).to be true
+        end
+
+        it "The winner? method should return false" do
```
Comment (whiz25) [inline spec/board_spec.rb:82]: - [x] Board Drawn game The winner? method should return false

File: spec/game_spec.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+        it "Retrieve player1's marker. Should be X" do
+            expect(game.player1.marker).to eq(marker1)
+        end
+
+        it "Retrieve player2's marker. Should be O" do
+            expect(game.player2.marker).to eq(marker2)
+        end
+
+        it "If new game is initialized the game state should be :active" do
```
Comment (whiz25) [inline spec/game_spec.rb:34]: - [x] Game Retrieve player names and markers If new game is initialized the game state should be :active

File: spec/game_spec.rb [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        end
+    end
+
+    describe "Game Won" do
+        let(:win) { ['X',2,3,4,'X',6,7,8,'X'] }
+        let(:current_player) { Player.new(player1, marker1) }
+        let(:board) { Board.new(win,9) }
+        subject(:game) { Game.new(player1, player2, current_player, board) }
+
+        it "Check if there is a winner. game_state should return :winner" do
```
Comment (whiz25) [inline spec/game_spec.rb:85]: - [x] Game Game Won Check if there is a winner. game_state should return :winner

File: spec/game_spec.rb [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        end
+    end
+
+    describe "Game Drawn" do
+        let(:drawn) { ['X','O','X','X','O','X','O','X','O'] }
+        let(:current_player) { Player.new(player1, marker1) }
+        let(:board) { Board.new(drawn,0) }
+        subject(:game) { Game.new(player1, player2, current_player, board) }
+
+        it "Check if the game is a draw. game_state should return :draw" do
```
Comment (whiz25) [inline spec/game_spec.rb:96]: - [x] Game Game Drawn Check if the game is a draw. game_state should return :draw

File: spec/board_spec.rb (hunk 5 above)
Comment (rsundar) [reply spec/board_spec.rb:53]: it does return true, that is the essence of the test.

Please clarify.

**Review by whiz25 — APPROVED** [review]
## Hey @rsundar 
Thanks for reaching out to me on slack. I was able to rectify the issue on my PC and now all tests pass as expected. So it's a :+1: from me. Merge and proceed to the next project.

### Martin Afani

**Comment by rsundar** [comment]
Thanks for all of your help and patience @maelfosso @whiz25 @wathigo. It's greatly appreciated :smile:

**Pull request merged**